File: Backend/SchoolManagement.Application/Services/TeacherService.cs
using AutoMapper;
using Microsoft.EntityFrameworkCore;
using SchoolManagement.Application.Interfaces;
using SchoolManagement.Domain.Entities;
using SchoolManagement.Domain.Enums;
using SchoolManagement.Infrastructure.Data;
using SchoolManagement.Shared.DTOs.Common;
using SchoolManagement.Shared.DTOs.Teachers;

namespace SchoolManagement.Application.Services;

public class TeacherService : ITeacherService
{
    private readonly SchoolDbContext _context;
    private readonly IMapper _mapper;

    public TeacherService(SchoolDbContext context, IMapper mapper)
    {
        _context = context;
        _mapper = mapper;
    }

    public async Task<PagedResult<TeacherDto>> GetAllTeachersAsync(int pageNumber, int pageSize)
    {
        var query = _context.Teachers
            .Include(t => t.User)
            .Where(t => !t.IsDeleted)
            .OrderBy(t => t.User.FirstName);

        var totalCount = await query.CountAsync();
        var teachers = await query
            .Skip((pageNumber - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        var teacherDtos = _mapper.Map<List<TeacherDto>>(teachers);

        return new PagedResult<TeacherDto>
        {
            Data = teacherDtos,
            TotalCount = totalCount,
            PageNumber = pageNumber,
            PageSize = pageSize
        };
    }

    public async Task<TeacherDto?> GetTeacherByIdAsync(int id)
    {
        var teacher = await _context.Teachers
            .Include(t => t.User)
            .FirstOrDefaultAsync(t => t.Id == id && !t.IsDeleted);

        return teacher != null ? _mapper.Map<TeacherDto>(teacher) : null;
    }

    public async Task<TeacherDto> CreateTeacherAsync(CreateTeacherDto createTeacherDto)
    {
        using var transaction = await _context.Database.BeginTransactionAsync();
        try
        {
            // Create User first
            var user = new User
            {
                FirstName = createTeacherDto.FirstName,
                LastName = createTeacherDto.LastName,
                Email = createTeacherDto.Email,
                PhoneNumber = createTeacherDto.PhoneNumber,
                DateOfBirth = createTeacherDto.DateOfBirth,
                Gender = createTeacherDto.Gender,
                Address = createTeacherDto.Address,
                Role = UserRole.Teacher,
                PasswordHash = BCrypt.Net.BCrypt.HashPassword("123456"), // Default password
                IsActive = true
            };

            _context.Users.Add(user);
            await _context.SaveChangesAsync();

            // Create Teacher
            var teacher = new Teacher
            {
                UserId = user.Id,
                EmployeeNumber = createTeacherDto.EmployeeNumber,
                HireDate = createTeacherDto.HireDate,
                Qualification = createTeacherDto.Qualification,
                Specialization = createTeacherDto.Specialization,
                BaseSalary = createTeacherDto.BaseSalary,
                SalaryType = createTeacherDto.SalaryType,
                HourlyRate = createTeacherDto.HourlyRate
            };

            _context.Teachers.Add(teacher);
            await _context.SaveChangesAsync();

            await transaction.CommitAsync();

            return await GetTeacherByIdAsync(teacher.Id) ?? throw new Exception("Failed to create teacher");
        }
        catch
        {
            await transaction.RollbackAsync();
            throw;
        }
    }

    public async Task<TeacherDto> UpdateTeacherAsync(int id, CreateTeacherDto updateTeacherDto)
    {
        var teacher = await _context.Teachers
            .Include(t => t.User)
            .FirstOrDefaultAsync(t => t.Id == id && !t.IsDeleted);

        if (teacher == null)
            throw new Exception("Teacher not found");

        // Update User
        teacher.User.FirstName = updateTeacherDto.FirstName;
        teacher.User.LastName = updateTeacherDto.LastName;
        teacher.User.Email = updateTeacherDto.Email;
        teacher.User.PhoneNumber = updateTeacherDto.PhoneNumber;
        teacher.User.DateOfBirth = updateTeacherDto.DateOfBirth;
        teacher.User.Gender = updateTeacherDto.Gender;
        teacher.User.Address = updateTeacherDto.Address;
        teacher.User.UpdatedAt = DateTime.UtcNow;

        // Update Teacher
        teacher.EmployeeNumber = updateTeacherDto.EmployeeNumber;
        teacher.HireDate = updateTeacherDto.HireDate;
        teacher.Qualification = updateTeacherDto.Qualification;
        teacher.Specialization = updateTeacherDto.Specialization;
        teacher.BaseSalary = updateTeacherDto.BaseSalary;
        teacher.SalaryType = updateTeacherDto.SalaryType;
        teacher.HourlyRate = updateTeacherDto.HourlyRate;
        teacher.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();

        return await GetTeacherByIdAsync(id) ?? throw new Exception("Failed to update teacher");
    }

    public async Task<bool> DeleteTeacherAsync(int id)
    {
        var teacher = await _context.Teachers
            .Include(t => t.User)
            .FirstOrDefaultAsync(t => t.Id == id && !t.IsDeleted);

        if (teacher == null)
            return false;

        teacher.IsDeleted = true;
        teacher.User.IsDeleted = true;
        teacher.UpdatedAt = DateTime.UtcNow;
        teacher.User.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<TeacherDto?> GetTeacherByUserIdAsync(int userId)
    {
        var teacher = await _context.Teachers
            .Include(t => t.User)
            .FirstOrDefaultAsync(t => t.UserId == userId && !t.IsDeleted);

        return teacher != null ? _mapper.Map<TeacherDto>(teacher) : null;
    }
}

File: Frontend/src/app/core/services/teacher.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { User } from '../models/user.model';

export interface Teacher {
  id: number;
  userId: number;
  employeeNumber: string;
  hireDate: Date;
  qualification: string;
  specialization: string;
  baseSalary: number;
  salaryType: number;
  hourlyRate: number;
  user: User;
  createdAt: Date;
  updatedAt?: Date;
}

export interface CreateTeacherRequest {
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber: string;
  dateOfBirth: Date;
  gender: number;
  address: string;
  employeeNumber: string;
  hireDate: Date;
  qualification: string;
  specialization: string;
  baseSalary: number;
  salaryType: number;
  hourlyRate: number;
}

export interface PagedResult<T> {
  data: T[];
  items: T[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
  totalPages: number;
  hasPreviousPage: boolean;
  hasNextPage: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class TeacherService {
  private readonly API_URL = `${environment.apiUrl}/teachers`;

  constructor(private http: HttpClient) {}

  getTeachers(pageNumber: number = 1, pageSize: number = 10): Observable<PagedResult<Teacher>> {
    const params = new HttpParams()
      .set('pageNumber', pageNumber.toString())
      .set('pageSize', pageSize.toString());
    
    return this.http.get<PagedResult<Teacher>>(this.API_URL, { params });
  }

  getTeacher(id: number): Observable<Teacher> {
    return this.http.get<Teacher>(`${this.API_URL}/${id}`);
  }

  createTeacher(teacher: CreateTeacherRequest): Observable<Teacher> {
    return this.http.post<Teacher>(this.API_URL, teacher);
  }

  updateTeacher(id: number, teacher: CreateTeacherRequest): Observable<Teacher> {
    return this.http.put<Teacher>(`${this.API_URL}/${id}`, teacher);
  }

  deleteTeacher(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }
}